### PR TITLE
cache: respond to watches in parallel v2

### PR DESCRIPTION
**Pull request by sschepens**
Supersedes #75 

Added concept of `ExecutorGroup` mostly based on netty's `EventExecutorGroup` for distributing `Executor`s to streams in a round robin fashion.

### CODE REVIEW

**Comment by sschepens** [comment]
tests are failing on circle but work ok on my computer, investigating

File: cache/src/main/java/io/envoyproxy/controlplane/cache/SimpleCache.java [hunk 1]
```diff
@@ -102,7 +114,7 @@ public Watch createWatch(
       Snapshot snapshot = snapshots.get(group);
       String version = snapshot == null ? "" : snapshot.version(request.getTypeUrl());
 
-      Watch watch = new Watch(ads, request, responseConsumer);
+      Watch watch = new Watch(ads, request, responseConsumer, executorGroup.next());
```
[the remark this answers is not in the document]
Comment (sschepens) [reply cache/src/main/java/io/envoyproxy/controlplane/cache/SimpleCache.java]: I was under the impression we created only one watch per stream, but this is only valid when not using ADS, we're probably going to need to move the RR to `DiscoveryServer`

**Comment by sschepens** [comment]
@snowp `DefaultExecutorGroup` is mostly copypasted from Netty `DefaultEventExecutorGroup`, wouldn't you think it's probably better to add a dependency on `netty-common` and use that?
We would avoid having to test and mantain this class.

**Comment by snowp** [comment]
I would actually go in the other direction: provide the minimum interface necessary here and a very simple implementation, maybe something that just returns the same fixed thread executor each time. That way you can use a netty implementation if you want without bringing all that complexity into the control plane API. 

Using a netty implementation would be as simple as doing 

```
  class NettyExecutorGroup implements ExecutorGroup {
    EventExecutorGroup nettyGroup;
    
    @Override public Executor next() {
      return nettyGroup.next()
  }
```

**Comment by sschepens** [comment]
@snowp thanks for the comments, totally agree.

I committed a different, simpler approach, moved `ExecutorGroup` and `DefaultExecutorGroup` to `server` to RR `Executor`s per stream.
Also `DefaultExecutorGroup` now defaults to the old behavior by just returning `directExecutor`.

This approach is much simpler and touches less things, please review.

File: server/src/main/java/io/envoyproxy/controlplane/server/DefaultExecutorGroup.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+/**
+ * Default implementation of {@link ExecutorGroup} which
+ * always returns {@link com.google.common.util.concurrent.MoreExecutors#directExecutor()}.
+ */
+public class DefaultExecutorGroup implements ExecutorGroup {
+  /**
+   * Creates a new instance.
+   */
+  public DefaultExecutorGroup() {
```
Comment (snowp) [inline server/src/main/java/io/envoyproxy/controlplane/server/DefaultExecutorGroup.java]: nit: no need for this default ctor

File: server/src/main/java/io/envoyproxy/controlplane/server/DefaultExecutorGroup.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package io.envoyproxy.controlplane.server;
+
+import com.google.common.util.concurrent.MoreExecutors;
+
+import java.util.concurrent.Executor;
+
+/**
+ * Default implementation of {@link ExecutorGroup} which
+ * always returns {@link com.google.common.util.concurrent.MoreExecutors#directExecutor()}.
```
Comment (snowp) [inline server/src/main/java/io/envoyproxy/controlplane/server/DefaultExecutorGroup.java]: supernit: shorten this to reference `MoreExecutors#directExecutor` instead of the fqn

File: server/src/main/java/io/envoyproxy/controlplane/server/ExecutorGroup.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+/**
+ * The {@link ExecutorGroup} is responsible for providing the {@link Executor}'s to use
+ * via its {@link #next()} method.
+ */
+public interface ExecutorGroup {
+  /**
+   * Return the next {@link Executor} to use.
+   *
+   */
+  Executor next();
```
[the remark this answers is not in the document]
Comment (sschepens) [reply server/src/main/java/io/envoyproxy/controlplane/server/ExecutorGroup.java]: I thought about this as well, but there's not much contextual information we can send at the moment

**Comment by sschepens** [comment]
> We don't have to solve this now, but it would be good to document this a bit more. I get the mechanics of it from a high-level, but I don't think I have a full picture of the different types of ExecutorGroup impls that people would want in the real world. Are there use-cases where you would want certain request streams to have an affinity for a particular Executor? Or is it mostly just direct executor vs. thread pool, with the pool size and pool entry chooser (e.g. netty's PowerOfTwoEventExecutorChooser) being the most interesting parts of the latter?

@joeyb I think it would mostly be just direct executor vs. thread pool, What I can see is people probably trying to reuse netty's worker groups, I would personally like that.
I can see little benefit from having stream affinity, even so, right now we have little to no context for taking these decisions.

**Comment by joeyb** [comment]
@sschepens - Cool, makes sense. If the need for more context comes up, then we can cross that bridge when we get there.

File: server/src/main/java/io/envoyproxy/controlplane/server/ExecutorGroup.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (snowp) [reply server/src/main/java/io/envoyproxy/controlplane/server/ExecutorGroup.java]: I think having each stream always using the same executor is good enough for now

**Comment by joeyb** [comment]
@sschepens, @snowp - FYI, I cut `0.1.12`, which includes this change.

**Comment by sschepens** [comment]
@joeyb thanks!
